source/server/connection_handler_impl.cc

```diff
@@ -54,6 +68,31 @@ void ConnectionHandlerImpl::removeListeners(uint64_t listener_tag) {
   }
 }
 
+void ConnectionHandlerImpl::removeFilterChains(
+    Network::DrainingFilterChains& draining_filter_chains, std::function<void()> completion) {
+  for (auto& listener : listeners_) {
+    // TODO(lambdai):
+    if (listener.second.listener_->listenerTag() ==
+        draining_filter_chains.getDrainingListenerTag()) {
+      ASSERT(!listener.second.tcp_listener_.has_value());
+      listener.second.tcp_listener_->get().removeFilterChains(
+          draining_filter_chains.getDrainingFilterChains());
+      Event::DeferredTaskUtil::deferredRun(dispatcher_, std::move(completion));
```

Comment:
What is the purpose of this deferred task?

Reply:
The `completion` will eventually trigger the config(filter chain and listener) destruction in master thread. This destruction need to happens after the connection/filterchain destruction. Since the connections are deferred deleted in worker thread, the completion should also deferred triggered. Also rely on the defer delete operation be FIFO